### ISSUE TITLE
NewPipeExtractor for searching, showing video details, showing comments, etc currently on YouTube, SoundClud (beta), media.ccc.de (beta), any PeerTube instance and Bandcamp

### ISSUE THREAD

**Issue by EmanuelLoos** [Schmiddiii/Tubefeeder#25]
Using [NewPipeExtractor](https://github.com/TeamNewPipe/NewPipeExtractor) Tubefeeder could:

- provide search functionality (in a language selected by the user)
- extract video details
- extract comments
- extract video in h264 and VP9
- extract audio in aac and opus
- display general info about channels
- search channels

NewPipe is a FOSS Android App that provides a FOSS, privacy respecting alternative to the official YouTube app. The NewPipeExtractor works a lot like the well known [youtube-dl](https://github.com/ytdl-org/youtube-dl) which mpv uses to extract videos, but is intended to extract all information needed to provide a full UI. Supporting both seems like a good idea to me and might help to enhance Tubefeeders functionality.

**Comment by Schmiddiii** [comment]
I already know about NewPipeExtractor, and already considered using it for the project.

The current main problem is that NewPipeExtractor is written in java, this project is written in rust.
It would definitly be possible to call java functions from rust, for example with [j4rs](https://crates.io/crates/j4rs).

Furthermore I do not know if basing on NewPipeExtractor would be even worth it. Are features like viewing comments even needed by the majority of the userbase?

As a frequent NewPipe-User I also experienced, that the app (or parts of it) just stops working every few months as YouTube changed something on their website and NewPipeExtractor needs to be fixed. They are usually very fast doing this, but there would still be some time between me noticing the app is not working and a fix be rolled out. In comparison to that the YouTube feed seems to always stay the same.

But if the userbase still wants these features, maybe the invidious api might be worth looking at.

All in all, I do not hate the idea of using NewPipeExtractor, but I think some discussion about whether the provided features are wanted is needed.

**Comment by EmanuelLoos** [comment]
> As a frequent NewPipe-User I also experienced, that the app (or parts of it) just stops working every few months as YouTube changed something on their website and NewPipeExtractor needs to be fixed. They are usually very fast doing this, but there would still be some time between me noticing the app is not working and a fix be rolled out. In comparison to that the YouTube feed seems to always stay the same.

Well youtube-dl is a much lager project and therefore gets fixes normally faster, but NewPipeExtractor could be used to provide these additional features in times it works.

> But if the userbase still wants these features, maybe the invidious api might be worth looking at.

I think that could be a good fallback, bit I'd choose a local software over a remote one. It could also be used to proxy videos.
Invidious has to extract comments and metadata or perform searches somehow. Maybe that code can be used better than NewPipeExtractor written in java.

**Comment by Schmiddiii** [comment]
> Well youtube-dl is a much lager project and therefore gets fixes normally faster, but NewPipeExtractor could be used to provide these additional features in times it works.

What do you mean by "could be used to provide these additional features in times it works"? Do you think this app should use NewPipe for searching, comments and other additional features, but keep the already working things? I do not think mixing NewPipe and code I wrote myself would be a great idea.

> I think that could be a good fallback, bit I'd choose a local software over a remote one.

I definitly agree that a local solution would be better.

**Comment by EmanuelLoos** [comment]
> What do you mean by "could be used to provide these additional features in times it works"? Do you think this app should use NewPipe for searching, comments and other additional features, but keep the already working things? I do not think mixing NewPipe and code I wrote myself would be a great idea.

I thought of supporting all four: your code, (external) youtube-dl, (external) NewPipeExtractor and the Invidious API
Not by mixing code, but by using external youtube-dl and external NewPipeExtractor if installed. How about calling them from command line?
That way if at least one of them works it works.
I don't know how difficult that would be though.

**Comment by Schmiddiii** [comment]
I do not think that would be feasable as I would need to support four different ways of interacting with youtube.
Furthermore if one of them would not work, it is mosk likely because YouTube changed something and all of them would not work.
NewPipeExtractor from the command-line would be a great idea, but I believe that such a thing does not exist.

**Comment by EmanuelLoos** [comment]
> Furthermore if one of them would not work, it is mosk likely because YouTube changed something and all of them would not work.

Invidious Instances could be fixed without the user having to do anything.

>  NewPipeExtractor from the command-line would be a great idea, but I believe that such a thing does not exist.

You might wnt to take a look at [np-cli](https://github.com/goofwear/np-cli).

**Comment by Schmiddiii** [comment]
> Invidious Instances could be fixed without the user having to do anything.

You are right about that.

> You might wnt to take a look at np-cli.

This seems to be only a demonstration, not a program that should be used productively. 
This is not a viable option.

**Comment by EmanuelLoos** [comment]
There are also the Extractors [FreeTube](https://github.com/FreeTubeApp) (JavaScript) uses, just to mention it.

**Comment by Schmiddiii** [comment]
Freetube uses Invidious.

I think that we should also consider a self-written but more powerful extractor, as I would also like to support lbry/odysee in the future and none of the above mentioned can do that.

**Comment by Schmiddiii** [comment]
I think I will have to move this into "not possible with my current knowlege", as:

# Why not use a existing extractor?

I would like to support lbry/odysee in the future. Currently I do not see any extractors that support youtube and odysee.
NewPipe may have it in some time, but I currently do not see any progress ([see here](https://github.com/TeamNewPipe/NewPipeExtractor/issues/434)).

# Why not write a extractor?

I have already tested around a little bit and stumbled across the problem, that youtube and lbry/odysee will need javascript to display their content. I do not see any possibility of running the javascript in a uncomplicated way, and even if it would be possible it would be a pretty large security concern.

# Why not use a extractor for youtube and the feed for lbry/odysee?

Mixing several different ways of retrieving content does not seem like a nice solution at all.

**Comment by EmanuelLoos** [comment]
> I have already tested around a little bit and stumbled across the problem, that youtube and lbry/odysee will need javascript to display their content.

Odysee is an online service which uses LBRY and LBRY is a censorship resistant content network with its index saved in the LBRY blockchain. I don't think you **need** JavaScript to display LBRY content, but unless you want to code an implementation of the LBRY network in another programing language you will be stuck with the implemantations that exist which are not all written in JavaScript. [Here is the lbry-sdk writen in Python](https://github.com/lbryio/lbry-sdk) [Here is lbry-android written in Java.](https://github.com/lbryio/lbry-android) ([also lbry-fdroid, the F-Droid version of the app](https://github.com/lbryio/lbry-fdroid)) [And finally lbry-desktop written in JavaScript.](https://github.com/lbryio/lbry-desktop)

**Comment by Schmiddiii** [comment]
After a little bit of searching around, I found [this](https://github.com/deep-gaurav/rusty_pipe), a youtube extractor written in Rust.
This seems to (somehow) extract everything wanted (see [this frontend](https://github.com/deep-gaurav/rusty_pipe_front)).

I will have to do more research but it might be worth looking at.

**Comment by marc0x1** [comment]
> After a little bit of searching around, I found [this](https://github.com/deep-gaurav/rusty_pipe), a youtube extractor written in Rust. This seems to (somehow) extract everything wanted (see [this frontend](https://github.com/deep-gaurav/rusty_pipe_front)).
> 
> I will have to do more research but it might be worth looking at.
that sounds interesting. Are there any new developments?

**Comment by Schmiddiii** [comment]
Tubefeeder is currently using the api of piped.kavin.rocks (or any other Piped instance). This API should already expose everything needed for a fully-featured client, but I currently don't have the time to do that and I am also not sure that I will ever do that as I like Tubefeeder as it currently is.

Note that I do not use `rusty_pipe` as this would require me to constantly update this library if Youtube decides to change anything about their website (which is pretty often if you look at NewPipe). I don't have the time or resources to do so.
